File: examples/buttons/server/rails/app/models/shop_user.rb
# frozen_string_literal: true

# A visitor, remembered by a signed cookie holding `id`. No email, no name, no
# password, no IP, no OAuth. A user with no credentials is the feature: the row
# exists so an order can outlive a browser session, and so the public feed has
# something anonymous to attribute a purchase to.
#
# TWO UUIDS, ON PURPOSE. `id` is the ownership token that travels in the signed
# cookie and is never rendered anywhere; `public_ref` is the handle the feed
# shows.
#
# Not because publishing `id` would be exploitable — the cookie is signed, so
# knowing its plaintext buys nothing. Because a published `id` stops being safe
# the moment ANYTHING else accepts a bare uuid: a debug parameter, an admin
# lookup, a well-meaning "simplification" to an unsigned cookie. One column
# removes the whole category, and this demo exists to be copied from.
class ShopUser < ApplicationRecord
  has_many :shop_orders, dependent: :destroy, inverse_of: :shop_user

  # How long a row may go untouched before `touch_seen!` writes again.
  SEEN_THROTTLE = 5.minutes

  # Throttled: a page load is a dozen requests and remembering "last seen" must
  # not be a write storm. `update_column` on purpose — no validations, no
  # callbacks, no updated_at churn for a timestamp nothing reads transactionally.
  def touch_seen!
    return if last_seen_at && last_seen_at > SEEN_THROTTLE.ago

    update_column(:last_seen_at, Time.current)
  end
end

File: examples/buttons/server/laravel/app/Models/ShopUser.php
<?php

declare(strict_types=1);

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;

/**
 * A visitor. Two uuids, two timestamps, no credentials.
 *
 * TWO UUIDS, ON PURPOSE: `id` is the ownership token that travels in the
 * encrypted cookie and is never rendered; `public_ref` is the handle the
 * recent-orders feed shows. A published `id` stops being safe the moment
 * anything else accepts a bare uuid.
 */
class ShopUser extends Model
{
    use HasUuids;

    /** How long a row may go untouched before the next page load writes again. */
    public const SEEN_THROTTLE_SECONDS = 5 * 60;

    protected $guarded = [];

    protected $casts = [
        'first_seen_at' => 'datetime',
        'last_seen_at' => 'datetime',
    ];

    public function uniqueIds(): array
    {
        return ['id', 'public_ref'];
    }

    public function orders(): HasMany
    {
        return $this->hasMany(ShopOrder::class, 'shop_user_id');
    }

    public function touchSeen(): void
    {
        if ($this->last_seen_at !== null && $this->last_seen_at->getTimestamp() > time() - self::SEEN_THROTTLE_SECONDS) {
            return;
        }
        $this->newQuery()->whereKey($this->getKey())->update(['last_seen_at' => now()]);
    }
}
